src/pf/optimize.lua

```diff
@@ -176,15 +176,24 @@ local function simplify(expr, is_tail)
       if type(lhs) == 'number' and type(rhs) == 'number' then
          return assert(folders[op])(lhs, rhs)
       elseif associative_binops[op] then
-         if type(rhs) == 'table' and rhs[1] == op and type(lhs) == 'number' then
+         -- Try to make the right operand a number.
+         if type(lhs) == 'number' then
             lhs, rhs = rhs, lhs
          end
-         if type(lhs) == 'table' and lhs[1] == op and type(rhs) == 'number' then
-            if type(lhs[2]) == 'number' then
-               return { op, assert(folders[op])(lhs[2], rhs), lhs[3] }
-            elseif type(lhs[3]) == 'number' then
+         if type(lhs) == 'table' and lhs[1] == op and type(lhs[3]) == 'number' then
+            if type(rhs) == 'number' then
+               -- (A op N1) op N2 -> A op (N1 op N2)
```

Comment:
I think it is probably sufficient to rewrite ((A op B) op C) as always A op (B op C).  Given that we always move numbers to the right this should push numbers down to the leaves where they can fold later.  Even this algorithm you have will require a fixpoint in ((A op N1) op (B op (C op N2))).  To do it in one pass would require a helper like simplify_if -- in this case simplify_assoc(op, A, simplify_assoc (op, N1, (B op (C op N2)))).  So IMO we either should do the minimal change and rely on the fixpoint, or do the comprehensive change and rely on a helper.  WDYT?